fix(request): generate correct redirect requests

edpaget opened the pull request:
Previously we could not generate correctly signed AuthnRequests using
HTTP-Redirect bindings. This happened for three reasons:

1. We were inserting the signature into the body of the SAML
AuthnRequest. When sending an AuthnRequest using HTTP-Redirect the
signature is included as a query param alongside the encoded SAML
request

2. We were specifying the wrong binding type. Our requests used
HTTP-Post binding. I think most SAML idps are pretty forgiving about
what a request looks like but get more strict when you put them into the
verification mode that checks signatures. Okta included here.

3. We did not specify a NameIDPolicy. Basically the same kind of thing
as point number 2.

As part of fixing those issues I decided to make our AuthnRequest flow
use more of the OpenSAML library instead of using the hiccup-based XML
generation. I plan to follow up to convert our Logout/SLO handling to
work in the same way and clean up unused functions in the crypto/coerce
namespaces once that is done.

edpaget committed:
fix(request): generate correct redirect requests

Previously we could not generate correctly signed AuthnRequests using
HTTP-Redirect bindings. This happened for three reasons:

1. We were inserting the signature into the body of the SAML
AuthnRequest. When sending an AuthnRequest using HTTP-Redirect the
signature is included as a query param alongside the encoded SAML
request

2. We were specifying the wrong binding type. Our requests used
HTTP-Post binding. I think most SAML idps are pretty forgiving about
what a request looks like but get more string when you put them into the
verification mode that checks signatures Okta included here.

3. We did not specify a NameIDPolicy. Basically the same kind of thing
as point number 2.

As part of fixing those issues I decided to make our AuthnRequest flow
use more of the OpenSAML library instead of using the hiccup-based XML
generation. I plan to follow up to convert our Logout/SLO handling to
work in the same way and clean up unused functions in the crypto/coerce
namespaces once that is done.

Closes #91

edpaget commented:
depends on @danielcompton's #90 I'll rebase away those commits once that's done.

escherize commented on src/saml20_clj/sp/request.clj:
```diff
@@ -3,9 +3,17 @@
+           [org.opensaml.saml.saml2.core.impl AuthnRequestBuilder IssuerBuilder NameIDPolicyBuilder]
+           org.opensaml.saml.saml2.metadata.impl.SingleSignOnServiceBuilder
+           org.opensaml.xmlsec.context.SecurityParametersContext
+           org.opensaml.xmlsec.SignatureSigningParameters))
```
We don't have the linter for this installed in this library, what do you think of adding 
`(set! *warn-on-reflection* true)` after the ns form? That would help us be more confident that we are doing the type hinting correctly*.

edpaget replied:
We have this linter run: https://github.com/metabase/saml20-clj/blob/master/.github/workflows/linters.yml#L40-L48 is that the same thing? Otherwise happy to do it.

escherize replied:
It is not blocking, but it would be a decent dev-chore to copy over some of our kondo config. In this case I was talking about the [warn-on-reflection](https://github.com/metabase/metabase/blob/master/.clj-kondo/config.edn#L33) linter.

edpaget committed:
fixup! set warn-on-reflection

edpaget committed:
Merge branch 'master' into ep-fix-redirect-generation

edpaget committed:
fix: generate signed SLO requests (#95)

* fix: generate signed SLO requests

Updates our SLO request generation system using the same principles as
the previous changes to the AuthnRequest generation.

This does drop the make-logout-request-xml function from the public API.

* Update src/saml20_clj/sp/request.clj

Co-authored-by: bryan <bryan.maass@gmail.com>

* Update src/saml20_clj/sp/request.clj

Co-authored-by: bryan <bryan.maass@gmail.com>

---------

Co-authored-by: bryan <bryan.maass@gmail.com>

edpaget committed:
feat: port metadata generation off of hiccup (#96)

* feat: port metadata generation off of hiccup

Allows us to fully remove hiccup XML generation for just using the
opensaml library and clean up other dead code and tests from using the
library directly.

Adds tests for metadata generation.

* fixup! add warn on reflection

edpaget committed:
feat: add support for logout responses (#97)

* feat: add support for logout responses

This adds a support for parsing out logout response from ring requests using the OpenSAML library and plugs them into the same validation framework we supply for other SAML responses.

* fixup! more complete support for all logout bindings

* fixup! reify java function interface

* fixup! deps formatting and docstring formatting

edpaget committed:
feat: update authnresponse to match logout (#99)

* feat: update authnresponse to match logout

Updates authnresponse to use the shared message validators that are also
used by the logout response. Update authnresponse to use the provided
POST and Redirect decoders from the OpenSAML libary.

This also makes the validators fail if data they need is not provided
rather than passing silently. Users of the library should disable
validators by passing in validator options rather than not passing
required data.

Updates the README to provide new directions for using the library.

The state manager and some of the validations are in a odd place.
OpenSAML provides state storage and management tools that we should
probably use instead since they will make it easier to construct
contexts to pass to built-in validators.

* fixup! export success predicate

* Update README.md

Co-authored-by: bryan <bryan.maass@gmail.com>

* feat: add a headless browser test for saml flow (#98)

* feat: add a headless browser test for saml flow

Just using keycloak for now

* chore: only publish when VERSION.txt changes (#101)

* fixup! update readme

---------

Co-authored-by: bryan <bryan.maass@gmail.com>

edpaget committed:
chore: update kondo config to match main metabase (#102)

Updates the kondo config to have the same base set of linters as the
main metabase repo. Fix warnings from the lints and remove some
additional unused code or move test-only code into the test ns.
